File: frontend/src/lib/oddsGaFallback.ts
import type { OddsJogadorEntry } from "@/types/dados";

/**
 * P(marcar ou assistir) assumindo independência entre G e A.
 * Entrada/saída em escala 0–100 (mesma das colunas G% e A%).
 */
export function calcularGaPctDeG_e_A(gPct: number, aPct: number): number {
  const pg = gPct / 100;
  const pa = aPct / 100;
  const pga = pg + pa - pg * pa;
  return Math.round(pga * 1000) / 10;
}

export function gaPctEfetivo(odds: OddsJogadorEntry | null | undefined): number | null {
  if (!odds) return null;
  if (odds.ga_pct != null) return odds.ga_pct;
  if (odds.g_pct != null && odds.a_pct != null) {
    return calcularGaPctDeG_e_A(odds.g_pct, odds.a_pct);
  }
  return null;
}

export function gaPctCalculado(odds: OddsJogadorEntry | null | undefined): boolean {
  return (
    odds != null &&
    odds.g_pct != null &&
    odds.a_pct != null &&
    odds.odds_ga == null
  );
}

/** Preenche GA% derivado de G% + A% quando o mercado não traz odd combinada. */
export function enriquecerOddsRodada(
  odds: OddsJogadorEntry | null | undefined,
): OddsJogadorEntry | null {
  if (!odds) return null;
  if (odds.ga_pct != null) return odds;
  const ga = gaPctEfetivo(odds);
  if (ga == null) return odds;
  return { ...odds, ga_pct: ga };
}

File: src/scrapers/odds_ga_fallback.py
"""GA% derivado de G% + A% (mesma fórmula do frontend)."""

from __future__ import annotations

from typing import Any


def calcular_ga_pct_de_g_a(g_pct: float, a_pct: float) -> float:
    """P(marcar ou assistir) assumindo independência entre G e A (escala 0–100)."""
    pg = g_pct / 100.0
    pa = a_pct / 100.0
    pga = pg + pa - pg * pa
    return round(pga * 100, 2)


def enriquecer_odds_entrada(entrada: dict[str, Any]) -> dict[str, Any]:
    """
    Preenche ga_pct quando a casa traz G% e A% mas não o mercado combinado.
    Não sobrescreve ga_pct já presente nem define odds_ga (indica derivação).
    """
    if entrada.get("ga_pct") is not None:
        return entrada
    g = entrada.get("g_pct")
    a = entrada.get("a_pct")
    if g is None or a is None:
        return entrada
    try:
        g_f = float(g)
        a_f = float(a)
    except (TypeError, ValueError):
        return entrada
    entrada["ga_pct"] = calcular_ga_pct_de_g_a(g_f, a_f)
    return entrada


def enriquecer_mapa_odds(odds: dict[str, dict]) -> dict[str, dict]:
    for entrada in odds.values():
        if isinstance(entrada, dict):
            enriquecer_odds_entrada(entrada)
    return odds
